engine/resources/docker.go

```diff
@@ -0,0 +1,415 @@
+)
+
+func init() {
+	engine.RegisterResource("docker", func() engine.Res { return &DockerRes{} })
+}
+
+// DockerRes is a docker account resource.
+type DockerRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+
+	State string                     `yaml:"state"` // running, stopped, or removed
+	Image string                     `yaml:"image"` // docker image or image:tag
+	Cmd   []string                   `yaml:"cmd"`   // command(s) to run on the container
+	Env   []string                   `yaml:"env"`   // environment variables ("VAR=val")
+	Ports map[string]map[int64]int64 `yaml:"ports"` // e.g. {"tcp" => {80 => 8080},},
+
+	client *client.Client // docker api client
+
+	init *engine.Init
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *DockerRes) Default() engine.Res {
+	return &DockerRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *DockerRes) Validate() error {
+	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout*time.Second)
+	defer cancel()
+
+	// validate state
+	if obj.State != containerRunning && obj.State != containerStopped && obj.State != containerRemoved {
+		return fmt.Errorf("state must be running stopped or removed")
+	}
+
+	// validate image
+	c, err := client.NewClient(client.DefaultDockerHost, dockerClientVersion, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+	resp, err := c.ImageSearch(ctx, obj.Image, types.ImageSearchOptions{Limit: 1})
+	if err != nil {
+		return fmt.Errorf("error searching for image: %s", err)
+	}
+	if len(resp) == 0 {
+		return fmt.Errorf("image: %s not found", obj.Image)
+	}
+	c.Close()
+
+	// validate env
+	for _, env := range obj.Env {
+		if !strings.Contains(env, "=") || strings.Contains(env, " ") {
+			return fmt.Errorf("invalid environment variable: %s", env)
+		}
+	}
+
+	// validate ports
+	for k, v := range obj.Ports {
+		if k != "tcp" && k != "udp" {
+			return fmt.Errorf("ports primary key should be tcp or udp")
+		}
+		for p, q := range v {
+			if (p < 1 || p > 0xFFFF) || (q < 1 || q > 0xFFFF) {
+				return fmt.Errorf("ports should be between 1 and 65535")
+			}
+		}
+	}
+	return nil
+}
+
+// Init runs some startup code for this resource.
+func (obj *DockerRes) Init(init *engine.Init) error {
+	var err error
+	obj.init = init // save for later
+
+	obj.client, err = client.NewClient(client.DefaultDockerHost, dockerClientVersion, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+
+	return nil
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *DockerRes) Close() error {
+	obj.client.Close()
+	return nil
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *DockerRes) Watch() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt)
```

Comment:
why is this here?

Reply:
To cancel the context in the event that we want to exit, and there are api requests still outstanding. Should I use a struct context and cancel it in close instead?